zcash_client_backend/src/fees/zip317.rs

```diff
@@ -0,0 +1,336 @@
+        Self { fee_rule }
+    }
+}
+
+impl ChangeStrategy for SingleOutputChangeStrategy {
+    type FeeRule = Zip317FeeRule;
+    type Error = Zip317FeeError;
+
+    fn fee_rule(&self) -> &Self::FeeRule {
+        &self.fee_rule
+    }
+
+    fn compute_balance<P: consensus::Parameters, NoteRefT: Clone>(
+        &self,
+        params: &P,
+        target_height: BlockHeight,
+        transparent_inputs: &[impl transparent::InputView],
+        transparent_outputs: &[impl transparent::OutputView],
+        sapling_inputs: &[impl sapling::InputView<NoteRefT>],
+        sapling_outputs: &[impl sapling::OutputView],
+        dust_output_policy: &DustOutputPolicy,
+    ) -> Result<TransactionBalance, ChangeError<Self::Error, NoteRefT>> {
+        let mut transparent_dust: Vec<_> = transparent_inputs
+            .iter()
+            .filter_map(|i| {
+                // for now, we're just assuming p2pkh inputs, so we don't check the size of the input
+                // script
+                if i.coin().value < self.fee_rule.marginal_fee() {
+                    Some(i.outpoint().clone())
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        let mut sapling_dust: Vec<_> = sapling_inputs
+            .iter()
+            .filter_map(|i| {
+                if i.value() < self.fee_rule.marginal_fee() {
+                    Some(i.note_id().clone())
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        // Depending on the shape of the transaction, we may be able to spend up to
+        // `grace_actions - 1` dust inputs. If we don't have any dust inputs though,
+        // we don't need to worry about any of that.
+        if !(transparent_dust.is_empty() && sapling_dust.is_empty()) {
+            let t_non_dust = transparent_inputs.len() - transparent_dust.len();
+            let t_allowed_dust = transparent_outputs.len().saturating_sub(t_non_dust);
+
+            // We add one to the sapling outputs for the (single) change output Note that this
+            // means that wallet-internal shielding transactions are an opportunity to spend a dust
+            // note.
+            let s_non_dust = sapling_inputs.len() - sapling_dust.len();
+            let s_allowed_dust = (sapling_outputs.len() + 1).saturating_sub(s_non_dust);
+
+            let available_grace_inputs = self
+                .fee_rule
+                .grace_actions()
+                .saturating_sub(t_non_dust)
+                .saturating_sub(s_non_dust);
+
+            let mut t_disallowed_dust = transparent_dust.len().saturating_sub(t_allowed_dust);
+            let mut s_disallowed_dust = sapling_dust.len().saturating_sub(s_allowed_dust);
+
+            if available_grace_inputs > 0 {
+                // if we have available grace inputs, allocate them first to transparent dust
+                // and then to sapling dust
```

Comment:
To clarify, the assumption here is that if any transparent inputs have been provided to this `ChangeStrategy`, the caller is okay with any of those inputs being spent, so we don't need to consider the privacy implications here. It might be useful to call these out in a comment.